docs/oppen_algorithm.md

```diff
@@ -0,0 +1,359 @@
+# Derek C. Oppen's pretty pritning algorithm
+
+Derek C. Oppen's algorithm provides an efficient solution for formatting and printing code with clean indentation that respects a specified screen width.
+
+When working with large blocks of code, ensuring proper indentation and alignment is crucial for readability and maintainability. However, achieving this while adhering to a specific line width can be a challenging task.
+
+This algorithm is great if you are working with code, documentation, logging, or any situation that requires pretty-printing.
+
+## The printing tokens
+
+Oppen's algorithm expects a list of tokens describing the structure of the expected output.
+
+The main tokens are:
+
+|   Name    | Description
+|-----------|------------
+| STRING    | String literal to be displayed.
+| BEGIN     | Opens a group.
+| END       | Closes a group.
+| BREAK     | Potential line break (adds a new line if the text doesn't fit on the line).
+| LINEBREAK | Guaranteed line break.
+
+A group can contain any number of groups, `String`, `Break`, and a `LineBreak`. Groups exist to containerize different _groups_ of tokens with respect to line breaking.
+
+Containerizing groups allows the creation of subsets that are independent from other subsets, and semi-independent from their parent subsets.
+
+Example for a maximum desired printing width of 3:
```

Comment:
Let's take an example to explain what we mean by containerizing groups with respect to line breaking.

Let's say we define our screen width to 3, and we want to print a simple group, with the following tokens: